eras/conway/impl/src/Cardano/Ledger/Conway/Rules/GovCert.hs

```diff
@@ -145,73 +146,59 @@ instance
-      -- TODO: check against a new PParam `drepDeposit`, once PParams are updated. -- someCheck ?! ConwayDRepIncorrectDeposit deposit
-      drepState <- do
-        expiryEpoch <-
-          calcDRepExpiryEpoch
-            cgceCurrentSlot
-            (cgcePParams ^. ppDRepActivityL)
-        pure
-          DRepState
-            { drepExpiry = expiryEpoch
-            , drepAnchor = mAnchor
-            }
-      pure $ vState {vsDReps = Map.insert cred drepState vsDReps}
-    ConwayUnRegDRep cred _deposit -> do
-      -- TODO: check against a new PParam `drepDeposit`, once PParams are updated. -- someCheck ?! ConwayDRepIncorrectDeposit deposit
+      deposit == ppDRepDeposit ?! ConwayDRepIncorrectDeposit deposit ppDRepDeposit
+      pure
+        vState
+          { vsDReps =
+              Map.insert cred (DRepState (cgceCurrentEpoch + ppDRepActivity) mAnchor ppDRepDeposit) vsDReps
+          }
+    ConwayUnRegDRep cred deposit -> do
       Map.member cred vsDReps ?! ConwayDRepNotRegistered cred
-      pure $ vState {vsDReps = Map.delete cred vsDReps}
+      checkDepositAgainstPaidDeposit vsDReps cred deposit
+      pure vState {vsDReps = Map.delete cred vsDReps}
     ConwayAuthCommitteeHotKey coldCred hotCred ->
-      overwriteCommitteeHotCred vState coldCred (Just hotCred)
+      checkAndOverwriteCommitteeHotCred vState coldCred $ Just hotCred
     ConwayResignCommitteeColdKey coldCred ->
-      overwriteCommitteeHotCred vState coldCred Nothing
+      checkAndOverwriteCommitteeHotCred vState coldCred Nothing
+    -- Update a DRep expiry too along with its anchor.
     ConwayUpdateDRep cred mAnchor -> do
-      Map.notMember cred vsDReps ?! ConwayDRepNotRegistered cred
-      expiryEpoch <-
-        calcDRepExpiryEpoch
-          cgceCurrentSlot
-          (cgcePParams ^. ppDRepActivityL)
-      let updateDRepState drepState =
-            drepState
-              & drepExpiryL .~ expiryEpoch
-              & drepAnchorL .~ mAnchor
-      pure $ vState & vsDRepsL %~ Map.update (Just . updateDRepState) cred
+      Map.member cred vsDReps ?! ConwayDRepNotRegistered cred
+      pure
+        vState
+          { vsDReps =
+              Map.adjust
+                (\drepState -> drepState & drepExpiryL .~ cgceCurrentEpoch + ppDRepActivity & drepAnchorL .~ mAnchor)
+                cred
+                vsDReps
+          }
   where
-    overwriteCommitteeHotCred vState coldCred hotCred = do
-      let hotCredMap = csCommitteeCreds $ vsCommitteeState vState
-      ((isNothing <$> Map.lookup coldCred hotCredMap) /= Just True)
-        ?! ConwayCommitteeHasResigned coldCred
-      pure $
+    checkColdCredHasNotResigned coldCred csCommitteeCreds =
+      ((isNothing <$> Map.lookup coldCred csCommitteeCreds) /= Just True)
+        ?! ConwayCommitteeHasPreviouslyResigned coldCred
+    checkDepositAgainstPaidDeposit vsDReps cred deposit =
+      forM_ (Map.lookup cred vsDReps) $ \drepState ->
+        let paidDeposit = drepState ^. drepDepositL
```

Comment:
```suggestion
      let drepState = Map.lookup cred vsDReps
          paidDeposit = drepState ^. drepDepositL
```

Reply:
`Map.lookup` gives us a `Maybe a`. Do you mean we should use the lens `_Just . _1 . drepDepositL` ?

Reply:
I'd rather use `do` notation here in that case, my first instinct when reading `forM_` was that it's traversing a list of some sort :grin:

Reply:
Something like
```haskell
do
  drepState Map.lookup cred vsDReps
  ...

Reply:
Do you mean I should add a `do` at the top of the function?

Reply:
I think it'd be even better to throw a different predicate failure if the lookup fails here

Reply:
How about I put the membership check predicate in the same function so they are inseparable when this function is called, like:

```haskell
    checkDepositAgainstPaidDeposit vsDReps cred deposit = do
      Map.member cred vsDReps ?! ConwayDRepNotRegistered cred
      forM_ (Map.lookup cred vsDReps) $ \drepState ->
        let paidDeposit = drepState ^. drepDepositL
         in deposit == paidDeposit ?! ConwayDRepIncorrectDeposit deposit paidDeposit
```

Reply:
ignore the `do` comment :)

Reply:
Yeah, I think that's a good idea

Reply:
but wait, why not just case match on the result of lookup?

Reply:
Yes, that's a better idea for sure :100:!

Reply:
getting a `Just x` from the lookup is a way better proof that the value is in the map than the `Bool` you get from `Map.member`